### ISSUE TITLE
Enhancement to starting multiple nodes with the same configuration

### ISSUE THREAD

**Issue by LittleJohnII** [wildfly-extras/sunstone#69]
It is currently possible to start a variable number of nodes with the same configuration with Sunstone, but it's a little clunky.

To do that, we define a node with complete configuration:

```
node.node1.stuff=stuff
node.node1.moreStuff=moreStuff
```

And then we add a second node with the template property:

```
node.node2.template=node1
```

We can also change any properties for `node2` that are different from the template settings. This approach can be chained (e.g. `node3` inherits from `node2`, `node2` from `node1`).

**Issues with the current approach:**

1. This configuration usually resides in the `sunstone.properties` file or an equivalent, because it can get large. It can, of course, be extracted and passed as a list of java system properties, but then, ...
1. The configuration requires at least one property per additional node. Starting a test with 40 servers means we'll be passing 40 properties.
1. The nodes are started by passing a list of node names (String) to the cloud provider. This means we need to know the node names before we start them. It is possible to establish a convention - e.g. if the `.properties` file related to the test defines `node1` as a template, we'll be looking for properties for `node2`, `node3`, etc. and we'll stop when we don't find any template property for `node4`. This is not an elegant solution though - it would be better if we could extract a list of node names from a property that has been passed by a user - e.g.:
```
node.node1.templateTo=node2,node3
```

**Proposed solution:**

Exactly what I used as an example above, we introduce property `node.<nodeName>.templateTo`, which will list the node names that will be automatically configured as if they were declared with `node.<newNodeName>.template`.

**Requirements:**

- The user needs a constant amount of properties (preferably one) to declare any number of additional nodes, if they all inherit from a single template.
- It is still possible to change any properties for the new nodes, just as if they were declared via `template`.
- Java system properties need to be recognized. This will allow us to change the number of nodes the test uses without touching the test or the `.properties` files.

**Non-requirements and notes:**

- Tests will still need to account for a variable number of nodes. There may or may not be API provided for easier manipulation (to be decided).
- Bare metal providers need a list of nodes when declared. This proposition makes that list variable. It would be nice if adding nodes would not require the user to redefine the list of nodes for bare metal providers. This may not be easy to implement and is only a little inconvenient, so it's not a requirement.

**Comment by kwart** [comment]
Proposal looks good.
It just has to be clear what happens when a node has configured more templates:
```properties
node.node1.templateTo=node2,node3
node.node2.template=node0
# or
node.nodeA.templateTo=nodeZ
node.nodeB.templateTo=nodeZ
```
IMO such configurations should not be allowed. Then a question comes, when this should be checked.

**Comment by LittleJohnII** [comment]
Agreed that such configuration should not be allowed. I'm thinking of implementing this in `CloudProperties`. Whenever the properties are loaded (either in `#reset` or `#load`), it'll go through the properties and for each node in `templateTo` it will add a reverse property `template`. E.g.:
```
node.node1.templateTo=node2,node3
node.node2.baremetal.publicAddress=...
```
will internally translate to
```
node.node1.templateTo=node2,node3
node.node2.template=node1
node.node2.baremetal.publicAddress=...
node.node3.template=node1
```
This has the advantage that the conflict can be detected and acted upon early (when reading properties). I'll just throw an exception listing the offending properties. Another advantage is we don't have to change anything beyond `CloudProperties` (most notably `ObjectProperties`, which operate with the `template` property, stay untouched).

The disadvantage is that the `CloudProperties` instance will contain properties which the user did not provide (the automatic `template` mappings) which might be confusing. I plan to document the behavior.

**Comment by LittleJohnII** [comment]
Resolved with #72
